### PR TITLE
Start zPages Experimental Spec

### PR DESCRIPTION
**Pull request by jajanet**
This PR begins providing more insight into what zPages are, their background, different types utilized in the past (including uses and other details), and other avenues to explore for zPages.

Implementation insights are also provided for TraceZ based off the findings from the OpenTelemetry C++ TraceZ project. Some links are provided, and more will be added in the future along with recommendations on interfacing more specifically using the OpenTelemetry API/SDK.

Related [OTEP](https://github.com/open-telemetry/oteps/blob/master/text/0110-z-pages.md)

@SergeyKanzhelev

### CODE REVIEW

File: experimental/zpages.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+While zPages are uniquely useful in being more lightweight and quicker compared to installing external exporters like Jaeger and Zipkin, they still offer many useful ways to debug and gain insight into applications. The uses depend on the type of zPage, which is detailed below.
+
+## Types of zPages
+### TraceZ
+TraceZ shows information on tracing, including aggregation counts for latency, running, and errors for spans grouped by name. It also allows users to look closer at details for spans that are sampled.
+
+This type of zPage is useful particularly for debugging latency issues (slow parts of applications), deadlocks (running spans that don't end), and errors (where error happen and what types). They're also good for spotting patterns by showing which latency speeds are typical for operations with a given span name.
```
Comment (anuraaga) [inline experimental/zpages.md:32]: Running spans that don't end are often instrumentation bugs I wouldn't define deadlock in terms of these.

File: experimental/zpages.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+Traditionally, zPages have approached this by rendering webpages purely on the server-side. This means the server would only serve statuc resources (HTML, CSS and possibly Javascript) when the user accesses a given endpoint. Based on the type of zPage and the server language used, a pure server-side approach would generate HTML pages using hardcoded strings from scratch or using a template. All zPages need some server-side rendering.
+
+Optionally, there could also be an API layer that translates native data structures to JSON strings for a frontend to use when designated endpoints are accessed. This API layer would be paired with a frontend that provides client-side functionality, which would need Javascript. The frontend Javascript would use the API by requesting information at endpoints to add updates to the HTML DOM without unnecessarily requesting and re-rendering static resources. This makes initial page loads quicker and requires no knowledge of client-side rendering.
+
+-------------
+## Future Possibilities
+### Out-process
+zPages that compatible across different languages, with processing of information happening outside of applications
```
Comment (anuraaga) [inline experimental/zpages.md:91]: Is this related to the zpage extension on the collector? I guess the otlp receiver should already have zpages similar to what the SDK provides?

File: experimental/zpages.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+> TODO
+
+## Shared zPages Components
+### Wrapper
+Each implementation ideally creates a wrapper class for zPages that allows users to add them all with minimal effort like an API, which could be as simple as adding 2 lines of code to include zPages. (importing zPages and initializing it).
+
+This wrapper class acts as an injection point, running all of the different zPages in a background thread without users needing knowledge of how they work under the hood. An example: if a language has TraceZ and RPCz implemented, then the wrapper will spin up servers for both when a user constructs a zPages class instance. Adding StatsZ would mean those developers would only add additional code to extract the needed information and display them on a page. 
+
+### HTTP Server
+All zPages have some sort of HTTP Server component to render their information on a webpage when a host:port and endpoint is accessed.
```
Comment (anuraaga) [inline experimental/zpages.md:82]: One thing I've been emphasizing in the Java implementation is being able to mount zpages in an existing server, for example by calling into the zpage logic from a servlet. This may be an interesting point to include in the spec.

File: experimental/zpages.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
+# zPages
```
[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply experimental/zpages.md:1]: I like z, because it doesn't have any connotation like `debug`. Debug is sometimes associated with something slow and not for production. 

Spring calls it "Production Ready Endpoints": https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-features.html#production-ready-endpoints Perhaps we can call it the same.

As for endpoint names, dropping `z` is OK with me.

[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply experimental/zpages.md:1]: @jajanet can you call the page "Production Ready Endpoints"? @yurishkuro @anuraaga @arminru ware you advocating for simply dropping z from the names of endpoints?

File: experimental/zpages.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+The idea of "zPages" originates from one of OpenTelemetry's predecessors, [OpenCensus](https://opencensus.io/). You can read more about zPages from the OpenCensus docs [here](https://opencensus.io/zpages) or the OTEP [here](https://github.com/open-telemetry/oteps/blob/master/text/0110-z-pages.md). OpenCensus has different zPage implementations in [Java](https://opencensus.io/zpages/java/), [Go](https://opencensus.io/zpages/go/), and [Node](https://opencensus.io/zpages/node/) and there has been similar internal solutions developed at companies like Uber. Within OpenTelemetry, zPages are also either developed or being developed in [C#](https://github.com/open-telemetry/opentelemetry-dotnet/tree/master/src/OpenTelemetry.Exporter.ZPages), Java, and C++.
+
+While zPages are uniquely useful in being more lightweight and quicker compared to installing external exporters like Jaeger and Zipkin, they still offer many useful ways to debug and gain insight into applications. The uses depend on the type of zPage, which is detailed below.
```
Comment (SergeyKanzhelev) [inline experimental/zpages.md:26]: Please mention that "For high throughput applications, external exporters are typically configured to send subset of telemetry for reach analysis to save costs, while zPages, being in-memory, can analyze more telemetry with the limited set of supported scenarios."

File: experimental/zpages.md (hunk 1 above)
Comment (SergeyKanzhelev) [reply experimental/zpages.md:32]:
```suggestion
This type of zPage is useful particularly for debugging latency issues (slow parts of applications), deadlocks and instrumentation problems (e.g. running spans that don't end), and errors (where error happen and what types). They're also good for spotting patterns by showing which latency speeds are typical for operations with a given span name.
```

File: experimental/zpages.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+## Types of zPages
+### TraceZ
+TraceZ shows information on tracing, including aggregation counts for latency, running, and errors for spans grouped by name. It also allows users to look closer at details for spans that are sampled.
+
+This type of zPage is useful particularly for debugging latency issues (slow parts of applications), deadlocks (running spans that don't end), and errors (where error happen and what types). They're also good for spotting patterns by showing which latency speeds are typical for operations with a given span name.
+
+You can read about TraceZ more [here](https://opencensus.io/zpages/java/#tracez).
```
[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply experimental/zpages.md:34]: +1

@jajanet if you want to move those docs here later, just add a TODO before this line.

File: experimental/zpages.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+### TraceConfigZ
+
+TraceConfigZ is closely related to TraceZ, allowing the user to modify how spans are sampled or how much data to keep in TraceZ by updating the TraceZ components accordingly.
```
Comment (SergeyKanzhelev) [inline experimental/zpages.md:38]: Are TraceConfig pages control sampling for spans that would be exported by regular exporter? I don't think TraceConfigZ is related to TraceZ.

File: experimental/zpages.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply experimental/zpages.md:1]: I don't consider "debug" to mean "slow". Debug means investigating performance of a process by looking deeper into the signals that are not available through normal monitoring. Changing logging level is also an example of debugging.

"Production ready" sounds much more narrow, like a healthcheck endpoint.

Another possible name is "introspection endpoints", however that is also more narrow than what the framework should provide.

File: experimental/zpages.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+### RPCz
+RPCz provides details on sent and received RPC messages, which is categorized  by RPC methods. This includes overall and error counts, average latency per call, RPCs sent per second, and input/output size per second.
+
+You can read about RPCz more [here](https://docs.google.com/document/d/1RWNyUIaKTYK12tck_rQjki4jTyHFfkD8sk54mjwRwso/edit#) and [here](https://opencensus.io/zpages/java/#rpcz).
+
+### StatsZ
+StatsZ is focused more on metrics, as it displays stats and measues for exported views. These views are grouped into directories using their namespaces
+
+You can read more about StatsZ [here](https://opencensus.io/zpages/java/#statsz)
```
Comment (SergeyKanzhelev) [inline experimental/zpages.md:52]: Same, add a TODO

File: experimental/zpages.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+## Design and Implementation
+### TraceZ Details
+To implement TraceZ, spans need to be collected, aggregated, and rendered on a webpage.
+
+For OpenTelemetry, a custom `span processor` can be made to interface with the `Tracer` API to collect spans. This span processor collects references to running spans and exports completed spans to its own memory or directly to an aggregator. An alternative to a span processor is using some sort of profiler.
+
+A `data aggregator` keeps a track of counts for running, error, and latency buckets for spans grouped by their name. It also samples some spans to provide users with more information. To prevent memory overload, only some spans are sampled for each bucket for each span name; for example, if that sampled span max number is set to 5, then only up to 55 pieces of span data can be kept for each span name in the aggregator (sampled_max * number of buckets = 5 * [running + error + 9 latency buckets] =  5 * 11 = 55).
```
Comment (SergeyKanzhelev) [inline experimental/zpages.md:60]: tracez is also designed to keep samples for every bucket of latency. This is a big benefit comparing to regular exporter. zPages will store unique spans that falls into specific latency bucket or having an error - those, that would likely be sampled out for the regular exporter

File: experimental/zpages.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (anuraaga) [reply experimental/zpages.md:1]: @SergeyKanzhelev I would drop the z from both the endpoints and names, I agree with @yurishkuro we can have a semantic name that is self-explanatory, and URLs with z's are not as clean as with normal English IMO.

debug seems fine to me, I might name a similar concept just "internal pages", or the other name suggestions seem fine to me since they're still self explanatory.

[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply experimental/zpages.md:1]: Would everybody be OK with "Introspection endpoints"? We might need more opinions, but I feel very strongly that "debug" will be considered as something not acceptable for production by customers.

File: experimental/zpages.md (hunk 1 above)
Comment (jajanet) [reply experimental/zpages.md]: Updated!

File: experimental/zpages.md (hunk 3 above)
Comment (jajanet) [reply experimental/zpages.md:77]: I added this information to line 87 at the end of the HTTP Server section:
> In either case, a benefit of reasoning about the zPages HTTP server as a separate component means that zPages can be mounted in a existing server. For example, this can be done in Java by calling this zPage logic from a servlet.

Let me know if it communicates the idea correctly!

File: experimental/zpages.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+For example, the sampling probability can be increased, decreased, or customized in other ways (i.e. depending on span parentage). Number of kept attritubtes, annotations, events, and links can also be adjusted.
+
+You can read about TraceConfigZ more [here](https://opencensus.io/zpages/java/#traceconfigz).
+
+### RPCz
+RPCz provides details on sent and received RPC messages, which is categorized  by RPC methods. This includes overall and error counts, average latency per call, RPCs sent per second, and input/output size per second.
+
+You can read about RPCz more [here](https://docs.google.com/document/d/1RWNyUIaKTYK12tck_rQjki4jTyHFfkD8sk54mjwRwso/edit#) and [here](https://opencensus.io/zpages/java/#rpcz).
+
+### StatsZ
```
[the remark this answers is not in the document]
Comment (jajanet) [reply experimental/zpages.md:43]: That makes sense! So everywhere with stats, I should replace with metrics?

File: experimental/zpages.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+### TraceConfigZ
+
+TraceConfigZ is closely related to TraceZ, allowing the user to modify how spans are sampled or how much data to keep in TraceZ by updating the TraceZ components accordingly.
+
+For example, the sampling probability can be increased, decreased, or customized in other ways (i.e. depending on span parentage). Number of kept attritubtes, annotations, events, and links can also be adjusted.
+
+You can read about TraceConfigZ more [here](https://opencensus.io/zpages/java/#traceconfigz).
+
+### RPCz
```
[the remark this answers is not in the document]
Comment (jajanet) [reply experimental/zpages.md:40]: From my understanding, metrics and collecting RPC call information is separate? I'm not exactly sure, I think @SergeyKanzhelev could provide better insight

File: experimental/zpages.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jajanet) [reply experimental/zpages.md:1]: From the conversation on naming conventions, my understanding is:

-  rename all references of zPages to Introspection Endpoints
-  drop the z from TraceZ, RPCz, etc

Is this correct?

File: experimental/zpages.md (hunk 5 above)
Comment (jajanet) [reply experimental/zpages.md]: Added to this line (26) as:
> For high throughput applications, zPages can also analyze more telemetry with the limited set of supported scenarios than external exporters; this is because zPages are in-memory while external exporters are typically configured to send subset of telemetry for reach analysis to save costs.

Let me know if I should change it!

File: experimental/zpages.md (hunk 9 above)
Comment (jajanet) [reply experimental/zpages.md]: Reclarified under line 30 (explanations/overviews on the types of zPages) as:
> In addition to these counts, TraceZ also keeps samples for for error, running, and each of the latency buckets for a given span names to allows users to look closer at span fields. This is particularlu useful compared to external exporters, which would otherwise likely sample them out.

Does that sound okay? I meant for this later section for outline more architecture details, but could restructure things as needed

File: experimental/zpages.md (hunk 7 above)
Comment (jajanet) [reply experimental/zpages.md]: I'm not sure; my understanding was that TraceConfigz is for TraceZ for OpenCensus, but I will look into it more and edit things accordingly. If it makes more sense, the decision here also doesn't need involve OpenCensus' zPages

File: experimental/zpages.md (hunk 8 above)
Comment (jajanet) [reply experimental/zpages.md]: Removed these links!

File: experimental/zpages.md (hunk 2 above)
Comment (jajanet) [reply experimental/zpages.md]: I didn't realize the collector had zPages, so I'm not sure how/if the collector zPages differs from the individual OTel language implementations of zPages. I will update you and change things once I look into it more, thanks for the information!

File: experimental/zpages.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jajanet) [reply experimental/zpages.md]: Removed these lines!

File: experimental/zpages.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (arminru) [reply experimental/zpages.md:1]: @jajanet Sounds good to me!
We could also add an _Internal_ or _OpenTelemetry SDK_ there to further disambiguate it from telemetry about user/app code.

[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply experimental/zpages.md:1]: I am late to the party, but perhaps "Diagnostic Pages" or "Diagnostic Endpoints"? I feel it conveys both the debugging and introspection meanings.

OpenCensus also uses the term for describing what it really is:
>They are useful to for in-process diagnostics

File: experimental/zpages.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+A `data aggregator` keeps track of counts for running, error, and latency buckets for spans grouped by their name. It also samples some spans to provide users with more information. To prevent memory overload, only some spans are sampled for each bucket for each span name; for example, if that sampled span max number is set to 5, then only up to 55 pieces of span data can be kept for each span name in the aggregator (sampled_max * number of buckets = 5 * [running + error + 9 latency buckets] =  5 * 11 = 55).
+
+When the user visits the TraceZ endpoint, likely something similar to `host:port/tracez`, then the distribution of latencies for span names will be rendered. When clicking on buckets counts for a span name, additional details on individual sampled spans for that bucket would be shown. These details would include trace ID, parent ID, span ID, start time, attributes, and more depending on the type of bucket (running, error, or latency) and what's implemented/recorded in the other components. See [HTTP Server](#http-server) for more information on implementation.
```
Comment (tigrannajaryan) [inline experimental/zpages.md:55]: Do we standardize the port number? It will make the discovery easier.

File: experimental/zpages.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+The UI/frontend/rendering layer is the HTML, CSS, and Javascript itself, in contrast to the logic to serve those files. This frontend uses the data layer's API on the client-side within the browser with Javascript by accessing certain endpoints depending on the user's actions. The data returned interacts with the Javascript, which determines and executes the logic necessary to render updates to the HTML DOM. Modifying the HTML DOM means there are no unnecessary requesting and re-rendering static files, and only parts of the webpage are changed. This makes subsequent data queries quicker and requires no knowledge of client-side rendering for the zPages developer.
```
Comment (tigrannajaryan) [inline experimental/zpages.md:85]: Does this imply that rendered zPages will be ONLY available if the client is JS-enabled? In troubleshooting scenarios I may have no access to a full-blown browser and just fire a wget against the zPage.

I agree that the separation to the data collection and rendering components is useful. I think it is also very useful to expose the zPage data purely as data, e.g. in JSON format, available via a REST API. However, I think there is also a value in simple human-readable HTML/CSS output that is rendered on server side and is available without requiring JS-enabled web clients.

I think a good approach would be to render on the server-side but for each endpoint also support JSON-encoded responses if the client requests it (via a request parameter or via "Accept" HTTP header).

Comment (SergeyKanzhelev) [reply experimental/zpages.md:85]: this is how OTEP is formulated and experimental code is implemented. Perhaps those endpoints are important to document. @jajanet can you please add a TODO in the document saying that those data endpoints needs to be documented as well?

File: experimental/zpages.md (hunk 12 above)
Comment (SergeyKanzhelev) [reply experimental/zpages.md:55]: not at the moment. I think it's not critical for the experimental stage unless we will get some usability results. Perhaps a separate research will be needed what port other things like springboot endpoints standardize on...

File: experimental/zpages.md (hunk 13 above)
Comment (jajanet) [reply experimental/zpages.md:85]: Yes, that's a good point and I'll do that! I'll also add points on security, as I didn't think about that before

File: experimental/zpages.md (hunk 7 above)
Comment (jajanet) [reply experimental/zpages.md]: I checked with someone who implemented TraceConfigz and they said TraceConfigz would effect all sampling for zPages and external backends. I updated the spec accordingly, and I'm wondering if that's that what we want in the spec?

**Comment by jajanet** [comment]
Hey everybody! Based on the discussion, let's have a poll on the new zPages name to help finalize this PR. Probably gonna have this up for at least a couple of days, and/or when there's a clear winner. In any case, the name discussion can be an open one and may be postponed to a later PR depending on the results. Vote using emojis! 

Prefix:
Laugh - Introspection
Hooray - Debugging
Heart - Diagnostic

Suffix:
Rocket - Endpoints
Eyes - Pages

@SergeyKanzhelev @anuraaga @yurishkuro @tigrannajaryan @arminru

**Review by carlosalberto — APPROVED** [review]

**Comment by SergeyKanzhelev** [comment]
this is experimental feature with the lower bar for approvals. Merging for quicker iterations...

**Comment by SergeyKanzhelev** [comment]
Naming thing should be addressed separately.

**Comment by bogdandrutu** [comment]
@jajanet @SergeyKanzhelev This PR broke the build, can you fix lint errors:
https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-specification/2038/workflows/bae573d8-2a3d-447f-98d5-0e28032a5621/jobs/3774
